components/raftstore/src/store/replication_mode.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use kvproto::metapb;
+use kvproto::replication_modepb::ReplicationStatus;
+use tikv_util::collections::{HashMap, HashMapEntry};
+
+/// A registry that maps store to a group.
+///
+/// A group ID is generated according to the label value. Stores with same
+/// label value will be mapped to the same group ID.
+#[derive(Default, Debug)]
+pub struct StoreGroup {
+    labels: HashMap<String, u64>,
+    stores: HashMap<u64, u64>,
+}
+
+impl StoreGroup {
+    /// Registers the store with given label value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the store is registered twice with different store ID.
+    pub fn register_store(&mut self, store_id: u64, label: String) -> u64 {
+        debug!("associated {} {}", store_id, label);
+        match self.stores.entry(store_id) {
+            HashMapEntry::Occupied(o) => {
+                assert_eq!(
+                    self.labels.get(&label),
+                    Some(o.get()),
+                    "store {:?}",
+                    store_id
+                );
+                *o.get()
+            }
+            HashMapEntry::Vacant(v) => {
+                let len = self.labels.len() as u64 + 1;
+                let id = *self.labels.entry(label).or_insert(len);
```

Comment:
Why the length of label can be defined as group id?

Reply:
Group ID is just an arbitrary number that maps to label value.